jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -30,11 +30,168 @@
+
+  switch (class_id)
+  {
+    case LIT_MAGIC_STRING_INT8_ARRAY_UL:
+    {
+      GET_ELEMENT (int8_t, src, ret_num);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT8_ARRAY_UL:
+    case LIT_MAGIC_STRING_UINT8_CLAMPED_ARRAY_UL:
+    {
+      GET_ELEMENT (uint8_t, src, ret_num);
+      break;
+    }
+    case LIT_MAGIC_STRING_INT16_ARRAY_UL:
+    {
+      GET_ELEMENT (int16_t, src, ret_num);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT16_ARRAY_UL:
+    {
+      GET_ELEMENT (uint16_t, src, ret_num);
+      break;
+    }
+    case LIT_MAGIC_STRING_INT32_ARRAY_UL:
+    {
+      GET_ELEMENT (int32_t, src, ret_num);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT32_ARRAY_UL:
+    {
+      GET_ELEMENT (uint32_t, src, ret_num);
+      break;
+    }
+    case LIT_MAGIC_STRING_FLOAT32_ARRAY_UL:
+    {
+      GET_ELEMENT (float, src, ret_num);
+      break;
+    }
+#if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64
+    case LIT_MAGIC_STRING_FLOAT64_ARRAY_UL:
+    {
+      GET_ELEMENT (double, src, ret_num);
+      break;
+    }
+#endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64 */
+    default:
+    {
+      JERRY_UNREACHABLE ();
+    }
+  }
+
+  return ret_num;
+} /* get_typedarray_element */
+
+/**
+ * set typedarray's element value
+ */
+static void
+set_typedarray_element (lit_utf8_byte_t *dst, /**< the location in the internal arraybuffer */
+                        ecma_number_t value, /**< the number value to set */
+                        lit_magic_string_id_t class_id) /**< class name of the typedarray */
+{
+#define SET_ELEMENT(type, location, number) \
+  int64_t v = (int64_t) number; \
+  *((type *) location) = (type) v;
+
+  switch (class_id)
+  {
+    case LIT_MAGIC_STRING_INT8_ARRAY_UL:
+    {
+      SET_ELEMENT (int8_t, dst, value);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT8_ARRAY_UL:
+    {
+      SET_ELEMENT (uint8_t, dst, value);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT8_CLAMPED_ARRAY_UL:
+    {
+      uint8_t clamped;
+
+      if (value > 255)
+      {
+        clamped = 255;
+      }
+      else if (value < 0)
+      {
+        clamped = 0;
+      }
+      else
+      {
+        clamped = (uint8_t) value;
+
+        if (clamped + 0.5 < value
+            || (clamped + 0.5 == value && clamped % 2 == 1))
+        {
+          clamped ++;
+        }
+      }
+      *((uint8_t *) dst) = clamped;
+      break;
+    }
+    case LIT_MAGIC_STRING_INT16_ARRAY_UL:
+    {
+      SET_ELEMENT (int16_t, dst, value);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT16_ARRAY_UL:
+    {
+      SET_ELEMENT (uint16_t, dst, value);
+      break;
+    }
+    case LIT_MAGIC_STRING_INT32_ARRAY_UL:
+    {
+      SET_ELEMENT (int32_t, dst, value);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT32_ARRAY_UL:
+    {
+      SET_ELEMENT (uint32_t, dst, value);
+      break;
+    }
+    case LIT_MAGIC_STRING_FLOAT32_ARRAY_UL:
+    {
+      *((float *) dst) = (float) value;
```

Comment:
for float/double type, it cannot use the SET_ELEMENT macro. 
SET_ELEMRNT macro will convert the ecma_number_t into a integer type, then convert to the target type. float -> int -> float will lose precision.

Why I have to convert to integer first, because "behavior of float type convert to shorter int type or unsigned int type is undefined"